crates/stc_ts_file_analyzer/src/analyzer/assign/mod.rs

```diff
@@ -2883,61 +2931,208 @@ impl Analyzer<'_, '_> {
+                                return Ok(());
+                            }
+                            // type A = Uppercase<`aB${string | number}`> - valid
+                            // type A = Uppercase<`aB${string | { x: string }}`>; - invalid
+                            let is_valid_union = union.types.iter().all(|v| {
+                                v.is_num_like()
+                                    || v.is_bigint_like()
+                                    || v.is_bool_like()
+                                    || v.is_str_like()
+                                    || v.is_null()
+                                    || v.is_undefined()
+                            });
+
+                            if !is_valid_union {
+                                return Err(ErrorKind::AssignFailed {
+                                    span: union.span(),
+                                    left: box Type::StringMapping(to.clone()),
+                                    right_ident: None,
+                                    right: box r.clone(),
+                                    cause: vec![],
+                                }
+                                .into());
+                            }
+                        }
+                        _ => {
+                            if !(v.is_any()
+                                || v.is_num_like()
+                                || v.is_bigint_like()
+                                || v.is_bool_like()
+                                || v.is_str_like()
+                                || v.is_null()
+                                || v.is_undefined())
+                            {
+                                return Err(ErrorKind::AssignFailed {
+                                    span: to.span(),
+                                    left: box Type::StringMapping(to.clone()),
+                                    right_ident: None,
+                                    right: box r.clone(),
+                                    cause: vec![],
+                                }
+                                .into());
+                            }
                         }
                     }
+                }
 
-                    return Err(ErrorKind::AssignFailed {
-                        span: r.span(),
-                        left: box Type::StringMapping(to.clone()),
-                        right_ident: None,
-                        right: box r.clone(),
-                        cause: vec![],
+                return Ok(());
+            }
+            Type::Alias(alias_ty) => {
+                return self.assign_to_intrinsic(&mut Default::default(), to, &alias_ty.ty, Default::default());
+            }
+            Type::Ref(ref_ty) => {
+                if let RTsEntityName::Ident(id) = &ref_ty.type_name {
+                    if let Ok(Some(ItemRef::Owned(v))) = &self.find_type(&id.into()) {
```

Comment:
`ItemRef` should never be matched with a pattern

Reply:
You can use `cloned().collect()` instead

Reply:
That makes sense but I don't think I understand where the `cloned.collect()` suggestion is supposed to fit in here :fearful:

Reply:
Removed this as well, it is never matched when running `./scripts/check.sh` and this way it won't be a false positive if we actually need to handle the `ref` match arm.

I believe I had added this when I was trying to make type alias validation work.